### PR TITLE
[RLlib] Increase backward compatibility of checkpoints.

### LINKED ISSUE #47426: [RLlib] Rename of `SingleAgentRLModuleSpec` to `RLModuleSpec` breaks restoring old checkpoints
### What happened + What you expected to happen

I wanted to restore checkpoints created with ray v2.34.0 with ray v2.35.0, which errors with

```
>>> from ray.rllib.algorithms import Algorithm
>>> Algorithm.from_checkpoint(path=".../ray_results/pbt_humanoid_test/PPO_Humanoid-v4_3338d_00003_3_2024-09-09_00-46-34/checkpoint_000014")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../.cache/pypoetry/virtualenvs/ray-f4XCQ9mO-py3.12/lib/python3.12/site-packages/ray/rllib/algorithms/algorithm.py", line 399, in from_checkpoint
    state = Algorithm._checkpoint_info_to_algorithm_state(
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../.cache/pypoetry/virtualenvs/ray-f4XCQ9mO-py3.12/lib/python3.12/site-packages/ray/rllib/algorithms/algorithm.py", line 3442, in _checkpoint_info_to_algorithm_state
    state = pickle.load(f)
            ^^^^^^^^^^^^^^
AttributeError: Can't get attribute 'SingleAgentRLModuleSpec' on <module 'ray.rllib.core.rl_module.rl_module' from '.../.cache/pypoetry/virtualenvs/ray-f4XCQ9mO-py3.12/lib/python3.12/site-packages/ray/rllib/core/rl_module/rl_module.py'>
```

I expected to be able to restore checkpoints from older ray versions after the upgrade.

Adding a line like
```py
SingleAgentRLModuleSpec = RLModuleSpec
```
to `ray/rllib/core/rl_module.rl_module.py` allows me to continue from the old checkpoint.

This seems to be caused by https://github.com/ray-project/ray/pull/46840

### Versions / Dependencies

python 3.10.12 / 3.12.5
ray v2.35.0

### Reproduction script

Create a checkpoint with ray 2.34.0, upgrade to ray 2.25.0 and try to restore the checkpoint

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

edit: replaced the traceback with a more minimal one.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

The `AlgorithmConfig` is still a pain point when trying to load older checkpoints (of the new stack specifically). The reason for this are usually attributes that were added between storing the checkpoint and loading it again (lately e.g. the `_torch_grad_scaler_class` attribute). This PR suggests a logic that enables loading older checkpoints with a newer version (of the new stack).

- The `AlgorithmConfig` class is provided by `set_state` and `from_state` to receive a dictionary state and initiate a config from it.
- The `Algorithm` does always create its `config` by calling the `from_state` method and thereby adds all attributes of the present `AlgorithmConfig` class version to it. 

## Related issue number
 Closes #47426 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/algorithm.py [hunk 1]
```diff
@@ -455,7 +455,7 @@ def __init__(
                 object. If unspecified, a default logger is created.
             **kwargs: Arguments passed to the Trainable base class.
         """
-        config = config or self.get_default_config()
+        config = config  # or self.get_default_config()
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:458]: nit: why remove the default config path?

File: rllib/algorithms/algorithm.py [hunk 2]
```diff
@@ -466,22 +466,31 @@ def __init__(
             # `self.get_default_config()` also returned a dict ->
             # Last resort: Create core AlgorithmConfig from merged dicts.
             if isinstance(default_config, dict):
-                config = AlgorithmConfig.from_dict(
-                    config_dict=self.merge_algorithm_configs(
-                        default_config, config, True
+                if "class" in config:
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:469]: remove this entire if-block: There are no more algos left that return a dict from their `get_default_config()` method.

File: rllib/algorithms/algorithm.py [hunk 3]
```diff
@@ -466,22 +466,31 @@ def __init__(
...
+                            default_config, config, True
+                        )
                     )
-                )
+
             # Default config is an AlgorithmConfig -> update its properties
             # from the given config dict.
             else:
-                config = default_config.update_from_dict(config)
+                if isinstance(config, dict) and "class" in config:
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:481]: Actually, let's do the following with this PR:
* Make `AlgorithmConfig` a `Checkpointable` and override the `get/set_state` and `get_ctor_args_and_kwargs` methods first:
```
def get_state(self, *, components=...):
    return [basically the AlgoConfig as a dict]

def set_state(self, state):
    # Here, we can conveniently check, whether keys in `state` exist as properties and override the correct properties as wanted.

def get_ctor_args_and_kwargs(self):
    return ()  # <- empty tuple AlgorithmConfigs are always constructed w/o any args/kwargs
```

Comment (sven1977) [reply rllib/algorithms/algorithm.py:481]: Then we can do this here:
```
if isinstance(config, dict):
    config = default_config
    config.set_state(config)
# else:
#     keep config as is (it's already a proper AlgorithmConfig object)
```

So, basically, we from here on treat user-provided config dicts as "AlgorithmConfig state dicts".

File: rllib/algorithms/algorithm_config.py [hunk 4]
```diff
@@ -781,6 +776,53 @@ def update_from_dict(
 
         return self
 
+    def get_state(self) -> Dict[str, Any]:
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:779]: Add `@override(Checkpointable)` and subclass AlgorithmConfig from `Checkpointable`.

File: rllib/algorithms/algorithm_config.py [hunk 5]
```diff
@@ -781,6 +776,53 @@ def update_from_dict(
...
+            for policy_id, policy_spec in state.pop("policies").items():
+                if isinstance(policy_spec, PolicySpec):
+                    policies_dict[policy_id] = policy_spec.get_state()
+                else:
+                    policies_dict[policy_id] = policy_spec
+            state["policies"] = policies_dict
+
+        return state
+
+    @classmethod
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:806]: Add `@override(Checkpointable)` and subclass AlgorithmConfig from `Checkpointable`.

Rename this method to `set_state()` and make it not a class method.

File: rllib/policy/policy.py [hunk 6]
```diff
@@ -124,6 +124,23 @@ def __eq__(self, other: "PolicySpec"):
             and self.config == other.config
         )
 
+    def get_state(self) -> Dict[str, Any]:
```
Comment (sven1977) [inline rllib/policy/policy.py:127]: this is ok. Leave as-is :)

File: rllib/algorithms/algorithm.py [hunk 7]
```diff
@@ -466,22 +466,31 @@ def __init__(
...
+                    config = default_config.from_state(config)
+                else:
+                    config = default_config.update_from_dict(config)
         else:
             default_config = self.get_default_config()
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:486]: Ah, I see. This answers my question above: We already do this `get_default_dict()` call here. Ok.

File: rllib/algorithms/algorithm.py [hunk 8]
```diff
@@ -466,22 +466,31 @@ def __init__(
...
             # Given AlgorithmConfig is not of the same type as the default config:
             # This could be the case e.g. if the user is building an algo from a
             # generic AlgorithmConfig() object.
             if not isinstance(config, type(default_config)):
                 config = default_config.update_from_dict(config.to_dict())
+            else:
+                config = default_config.from_state(config.get_state())
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:493]:
```suggestion
            default_config = self.get_default_config()
            config_state = config.get_state()
            config = default_config
            config.set_state(config_state)
```

Comment (sven1977) [reply rllib/algorithms/algorithm.py:493]: I think we can simplify here to the above suggestion ^

File: rllib/algorithms/algorithm.py [hunk 9]
```diff
@@ -2899,7 +2908,7 @@ def get_checkpointable_components(self) -> List[Tuple[str, "Checkpointable"]]:
     @override(Checkpointable)
     def get_ctor_args_and_kwargs(self) -> Tuple[Tuple, Dict[str, Any]]:
         return (
-            (self.config,),  # *args,
+            (self.config.get_state(),),  # *args,
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:2911]: Perfect!

Comment (sven1977) [reply rllib/algorithms/algorithm.py:2911]: Btw, we should probably do the same everywhere else `config` is part of the c'tor args/kwargs: Learner, LearnerGroup, and EnvRunner.

Comment (sven1977) [reply rllib/algorithms/algorithm.py:2911]: Then, we also have to make sure their c'tors also accept config states(!), not just AlgorithmConfig objects (the same as how Algorithm does it now).

**Review by sven1977 — COMMENTED** [review]
I really like this PR. It solves so many problems at once!

A few design change requests and nits, but overall already in very good shape! Thanks @simonsays1980 .

File: rllib/algorithms/algorithm.py (hunk 1 above)
Comment (simonsays1980) [reply rllib/algorithms/algorithm.py:458]: I see. Yes, this is impüortant when a user does not pass in any `config`. I will change this. Good catch!

File: rllib/algorithms/algorithm.py (hunk 7 above)
Comment (simonsays1980) [reply rllib/algorithms/algorithm.py:486]: Haha, now I haven't overviewed the whole logic myself. So, yes, here we set the default, if nothing else is provided.

File: rllib/algorithms/algorithm_config.py (hunk 4 above)
Comment (simonsays1980) [reply rllib/algorithms/algorithm_config.py:779]: Yeah. This looks sound now. I am so glad when the old stack is gone. Everything will become so clean then.

File: rllib/policy/policy.py (hunk 6 above)
Comment (simonsays1980) [reply rllib/policy/policy.py:127]: Puh! :D

**Comment by sven1977** [comment]
Oh, sorry, another thing. Can we also get rid of (rename):
* `AlgorithmConfig.to_dict()` -> `get_state()`
* `AlgorithmConfig.update_from_dict()` -> `set_state()`

Does this make sense? ^

**Review by sven1977 — APPROVED** [review]
LGTM now! Thanks @simonsays1980 .

**Pull request merged**